File: palindrome-partitioning.cc
// Palindrome Partitioning
#define ROF(i, a, b) for (int i = (b); --i >= (a); )
#define FOR(i, a, b) for (int i = (a); i < (b); i++)
#define REP(i, n) for (int i = 0; i < (n); i++)

class Solution {
private:
  vector<vector<bool>> f;
  vector<string> rr;
  vector<vector<string>> r;
  void g(string &s, int i) {
    if (i == s.size())
      r.push_back(rr);
    else
      REP(j, s.size())
        if (f[i][j]) {
          rr.push_back(s.substr(i, j-i+1));
          g(s, j+1);
          rr.pop_back();
        }
  }
public:
  vector<vector<string>> partition(string s) {
    int n = s.size();
    f.assign(n, vector<bool>(n));
    ROF(i, 0, n) {
      f[i][i] = true;
      if (i+1 < n && s[i] == s[i+1])
        f[i][i+1] = true;
      FOR(j, i+2, n)
        f[i][j] = f[i+1][j-1] && s[i] == s[j];
    }
    r.clear();
    g(s, 0);
    return r;
  }
};

///

#define REP1(i, n) for (remove_cv<remove_reference<decltype(n)>::type>::type i = 1; i <= (n); i++)

class Solution {
private:
  vector<string> r;
  vector<vector<string>> ret;
  string s;
  vector<int> r0, r1;
  void dfs(int i) {
    if (i == s.size()) {
      ret.push_back(r);
      return;
    }
    REP1(j, s.size()-i)
      if ((j%2 ? r1 : r0)[i+j/2] >= (j+1)/2) {
        r.push_back(s.substr(i, j));
        dfs(i+j);
        r.pop_back();
      }
  }
public:
  vector<vector<string>> partition(string s) {
    int n = s.size();
    r0.assign(n, 0);
    r1.assign(n, 0);
    // r1[i]: 以i爲中心的奇數長度最長迴文子串的長度爲2*r1[i]-1，延伸到i+r1[i]-1
    for (int f, g = 0, i = 0; i < n; i++)
      if (i < g && r1[2*f-i] != g-i)
        r1[i] = min(r1[2*f-i], g-i);
      else {
        f = i;
        g = max(g, i);
        for (; g < n && 2*f >= g && s[2*f-g] == s[g]; g++);
        r1[i] = g-f;
      }
    // r0[i]: 以i-1和i和i爲中心的偶數長度最長迴文子串的長度爲2*r0[i]，延伸到i+r0[i]-1
    for (int f, g = 0, i = 0; i < n; i++)
      if (i < g && r0[2*f-i] != g-i)
        r0[i] = min(r0[2*f-i], g-i);
      else {
        f = i;
        g = max(g, i);
        for (; g < n && 2*f-1 >= g && s[2*f-1-g] == s[g]; g++);
        r0[i] = g-f;
      }
    this->s = std::move(s);
    ret.clear();
    dfs(0);
    return ret;
  }
};

File: palindrome-partitioning.rs
impl Solution {
    pub fn partition(s: String) -> Vec<Vec<String>> {
        #[derive(Default)]
        struct C<'a> { s: &'a str, r0: Vec<usize>, r1: Vec<usize>, ret: Vec<Vec<String>>, r: Vec<String> }
        impl<'a> C<'a> {
            fn dfs(&mut self, i: usize) {
                if i == self.s.len() {
                    self.ret.push(self.r.clone());
                    return;
                }
                for j in 1..=self.s.len()-i {
                    if (if j%2!=0 { &self.r1 } else { &self.r0 })[i+j/2] >= (j+1)/2 {
                        self.r.push(self.s[i..i+j].to_string());
                        self.dfs(i+j);
                        self.r.pop();
                    }
                }
            }
        }

        let a = s.as_bytes(); let n = a.len();
        let mut f = 0; let mut g = 0;
        let mut c = C { s: &s, r0: vec![0; n], r1: vec![0; n], ..Default::default() };
        for i in 0..n {
            c.r1[i] = if i < g && c.r1[2*f-i] != g-i { c.r1[2*f-i].min(g-i) } else {
                f = i;
                g = g.max(i);
                while g < n && 2*f >= g && a[2*f-g] == a[g] { g += 1; }
                g-f
            }
        }
        g = 0;
        for i in 0..n {
            c.r0[i] = if i < g && c.r0[2*f-i] != g-i { c.r0[2*f-i].min(g-i) } else {
                f = i;
                g = g.max(i);
                while g < n && 2*f >= g+1 && a[2*f-1-g] == a[g] { g += 1; }
                g-f
            }
        }
        c.dfs(0);
        c.ret
    }
}

///

impl Solution {
    pub fn partition(s: String) -> Vec<Vec<String>> {
        fn dfs<'a,'b>(s: &'a str, path: &'a [Vec<usize>], i: usize, ret: &'b mut Vec<Vec<String>>, r: &'b mut Vec<&'a str>) {
            if i == s.len() {
                ret.push(r.iter().map(|x| x.to_string()).collect());
                return;
            }
            for &j in path[i].iter() {
                r.push(&s[i..j]);
                dfs(s, path, j, ret, r);
                r.pop();
            }
        }

        let a = s.as_bytes(); let n = a.len();
        let mut f = 0; let mut g = 0;
        let mut r0 = vec![0; n];
        let mut r1 = vec![0; n];
        let mut path = vec![vec![]; n];
        for i in 0..n {
            r1[i] = if i < g && r1[2*f-i] != g-i { r1[2*f-i].min(g-i) } else {
                f = i;
                g = g.max(i);
                while g < n && 2*f >= g && a[2*f-g] == a[g] { g += 1; }
                g-f
            };
            for j in 0..r1[i] {
                path[i-j].push(i+j+1);
            }
        }
        g = 0;
        for i in 0..n {
            r0[i] = if i < g && r0[2*f-i] != g-i { r0[2*f-i].min(g-i) } else {
                f = i;
                g = g.max(i);
                while g+1 < n && 2*f >= g && a[2*f-g] == a[g+1] { g += 1; }
                g-f
            };
            for j in 0..r0[i] {
                path[i-j].push(i+j+2);
            }
        }
        let mut ret = vec![];
        let mut r = vec![];
        dfs(&s, &path, 0, &mut ret, &mut r);
        ret
    }
}
